### PR TITLE
Use default credential chain for CLI auth and policies attached to nodes for Operator/API auth

### LINKED ISSUE #1134: Support aws_session_token for CLI auth
#### Description

In order to authenticate with the Cortex operator, the Cortex CLI should be able to use `aws_session_token` (currently only static credentials are supported).

Also, consider enabling auth via IAM role (e.g. inherited from Lambda, EC2)

### LINKED ISSUE #1865: CLI Auth
Authenticate requests to the operator from CLI/Python.

- [ ] bring back https (insecure)
- [ ] use aws credentials via the default toolchain (don't save credentials), (e.g. AWS_PROFILE=admin && cortex get)
- [ ] encode credentials, attach to a header and send to operator
- [ ] bring back IAM auth check

### PR DESCRIPTION
**Pull request by vishalbollu**
closes #1865, closes #1134

---

checklist:

- [x] run `make test` and `make lint`
- [x] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)
- [x] update docs and add any new files to `summary.md` (view in [gitbook](https://cortex-labs.gitbook.io/staging/-MOmCGMADSRNQahK3Kox/) after merging)

### CODE REVIEW

File: pkg/lib/aws/aws.go [hunk 1]
```diff
@@ -35,21 +35,12 @@ type Client struct {
...
-func NewFromCreds(region string, accessKeyID string, secretAccessKey string) (*Client, error) {
-	creds := credentials.NewStaticCredentials(accessKeyID, secretAccessKey, "")
-	return New(region, creds)
-}
-
 func NewFromClientS3Path(s3Path string, awsClient *Client) (*Client, error) {
 	if !awsClient.IsAnonymous {
 		if awsClient.AccessKeyID() == nil || awsClient.SecretAccessKey() == nil {
 			return nil, ErrorUnexpectedMissingCredentials(awsClient.AccessKeyID(), awsClient.SecretAccessKey())
 		}
```
Comment (deliahu) [inline pkg/lib/aws/aws.go:42]: Delete these lines

File: pkg/lib/aws/aws.go [hunk 2]
```diff
@@ -60,49 +51,65 @@ func NewFromClientS3Path(s3Path string, awsClient *Client) (*Client, error) {
...
-	region, err := GetBucketRegion(bucket)
+	if sess.Config.Credentials == nil {
+		return nil, ErrorUnableToFindCredentials()
+	}
+
+	_, err = sess.Config.Credentials.Get()
 	if err != nil {
-		return nil, err
+		return nil, ErrorUnableToFindCredentials()
 	}
```
Comment (deliahu) [inline pkg/lib/aws/aws.go:89]: Check for access key ID and secret (and that it works with e.g. ec2 IAM roles) (also below)

File: pkg/operator/endpoints/errors.go [hunk 3]
```diff
@@ -65,6 +67,20 @@ func ErrorAuthAPIError() error {
 	})
 }
 
+func ErrorAuthInvalid() error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrAuthInvalid,
+		Message: "invalid AWS credentials; run `cortex env configure` to configure your environment with credentials for any IAM user in the same AWS account as your cluster",
```
Comment (deliahu) [inline pkg/operator/endpoints/errors.go:73]: update error message

File: pkg/operator/endpoints/errors.go [hunk 4]
```diff
@@ -65,6 +67,20 @@ func ErrorAuthAPIError() error {
...
+	})
+}
+
+func ErrorAuthOtherAccount() error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrAuthOtherAccount,
+		Message: "the AWS account associated with your CLI's AWS credentials differs from the AWS account associated with your cluster's AWS credentials; run `cortex env configure` to configure your environment with credentials for any IAM user in the same AWS account as your cluster",
```
Comment (deliahu) [inline pkg/operator/endpoints/errors.go:80]: update error message

File: pkg/operator/endpoints/middleware.go [hunk 5]
```diff
@@ -58,6 +60,38 @@ func ClientIDMiddleware(next http.Handler) http.Handler {
...
+func AuthMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		authHeader := r.Header.Get("Authorization")
+
+		if authHeader == "" {
+			respondError(w, r, ErrorHeaderMissing("Authorization"))
+			return
+		}
+
+		if config.Provider == types.AWSProviderType {
```
Comment (deliahu) [inline pkg/operator/endpoints/middleware.go:72]: remove this if

File: pkg/operator/endpoints/middleware.go [hunk 6]
```diff
@@ -58,6 +60,38 @@ func ClientIDMiddleware(next http.Handler) http.Handler {
 	})
 }
 
+func AuthMiddleware(next http.Handler) http.Handler {
```
Comment (deliahu) [inline pkg/operator/endpoints/middleware.go:63]: rename to AWSAuthMiddleware

File: pkg/types/clusterconfig/cluster_config_aws.go [hunk 7]
```diff
@@ -42,6 +44,7 @@ import (
 const (
 	// the s3 url should be used (rather than the cloudfront URL) to avoid caching
 	_cniSupportedInstancesURL = "https://cortex-public.s3-us-west-2.amazonaws.com/cli-assets/cni_supported_instances.txt"
+	_defaultIAMPolicy         = "arn:aws:iam::aws:policy/AmazonS3FullAccess"
```
Comment (deliahu) [inline pkg/types/clusterconfig/cluster_config_aws.go:47]: nit: make this a slice, e.g. `_defaultIAMPolicies`

File: pkg/types/clusterconfig/cluster_config_aws.go [hunk 8]
```diff
@@ -92,6 +95,8 @@ type ManagedConfig struct {
 	SpotConfig                 *SpotConfig        `json:"spot_config" yaml:"spot_config"`
 	AvailabilityZones          []string           `json:"availability_zones" yaml:"availability_zones"`
 	SSLCertificateARN          *string            `json:"ssl_certificate_arn,omitempty" yaml:"ssl_certificate_arn,omitempty"`
+	CortexPolicyARN            string             `json:"cortex_policy_arn" yaml:"cortex_policy_arn"`
```
Comment (deliahu) [inline pkg/types/clusterconfig/cluster_config_aws.go:98]: add a comment that this is not user facing

File: pkg/types/clusterconfig/cluster_config_aws.go [hunk 9]
```diff
@@ -402,6 +407,22 @@ var ManagedConfigStructFieldValidations = []*cr.StructFieldValidation{
...
+			AllowEmpty:       true,
+			TreatNullAsEmpty: true,
+		},
+	},
+	{
+		StructField: "IAMPolicyARNs",
+		StringListValidation: &cr.StringListValidation{
+			Default:           []string{_defaultIAMPolicy},
+			AllowEmpty:        true,
+			AllowExplicitNull: true,
```
Comment (deliahu) [inline pkg/types/clusterconfig/cluster_config_aws.go:423]: Confirm that null will not cause any panics / exceptions

File: pkg/types/clusterconfig/cluster_config_aws.go [hunk 10]
```diff
@@ -1423,6 +1461,13 @@ func (mc *ManagedConfig) TelemetryEvent() map[string]interface{} {
 		event["ssl_certificate_arn._is_defined"] = true
 	}
 
+	// CortexPolicyARN should be managed by cortex
+
+	if !strset.New(_defaultIAMPolicy).IsEqual(strset.New(mc.IAMPolicyARNs...)) {
+		event["iam_policy_arns._is_defined"] = true
```
Comment (deliahu) [inline pkg/types/clusterconfig/cluster_config_aws.go:1467]: nit: `_is_custom`

File: pkg/types/clusterconfig/aws_policy.go [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Action": [
+                "sts:GetCallerIdentity",
+                "ecr:GetAuthorizationToken",
+                "ecr:BatchGetImage",
+                "elasticloadbalancing:Describe*",
+				"sqs:ListQueues"
```
Comment (deliahu) [inline pkg/types/clusterconfig/aws_policy.go:49]: convert indentation

File: pkg/types/clusterconfig/aws_policy.go [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+var _cortexPolicy = `
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Action": [
+                "sts:GetCallerIdentity",
+                "ecr:GetAuthorizationToken",
+                "ecr:BatchGetImage",
+                "elasticloadbalancing:Describe*",
```
Comment (deliahu) [inline pkg/types/clusterconfig/aws_policy.go:48]: This may not be necessary

File: pkg/types/clusterconfig/aws_policy.go [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+				"sqs:ListQueues"
+            ],
+            "Effect": "Allow",
+            "Resource": "*"
+        },
+		{
```
Comment (deliahu) [inline pkg/types/clusterconfig/aws_policy.go:54]: nit: indentation

File: pkg/types/clusterconfig/aws_policy.go [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+        },
+        {
+            "Effect": "Allow",
+            "Action": "s3:*",
+            "Resource": "arn:aws:s3:::{{ .Bucket }}"
+        },
+        {
+            "Effect": "Allow",
+            "Action": "logs:PutLogEvents",
+            "Resource": "arn:aws:logs:{{ .Region }}:{{ .AccountID }}:log-group:{{ .LogGroup }}:*:*"
```
Comment (deliahu) [inline pkg/types/clusterconfig/aws_policy.go:67]: might be able to be deleted

File: pkg/types/clusterconfig/aws_policy.go [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+func AddNewPolicyVersion(awsClient *aws.Client, policyARN string, policyDocument string) error {
+	policies, err := awsClient.IAM().ListPolicyVersions(&iam.ListPolicyVersionsInput{
+		PolicyArn: &policyARN,
+	})
+	if err != nil {
+		return err
+	}
+
+	if len(policies.Versions) == 0 {
+		return nil
```
Comment (deliahu) [inline pkg/types/clusterconfig/aws_policy.go:151]: consider returning an error here

File: pkg/cortex/serve/cortex_internal/lib/storage/s3.py [hunk 16]
```diff
@@ -37,10 +37,7 @@ def __init__(self, bucket=None, region=None, client_config={}):
         if region is not None:
             client_config["region_name"] = region
 
-        if not (os.getenv("AWS_ACCESS_KEY_ID") and os.getenv("AWS_SECRET_ACCESS_KEY")):
```
[the remark this answers is not in the document]
Comment (vishalbollu) [reply pkg/cortex/serve/cortex_internal/lib/storage/s3.py:40]: I completely missed this! For some reason I thought this code was for local. Thanks.

**Review by deliahu — APPROVED** [review]

**Pull request merged**